### PR TITLE
sokol_app.h win32: implements optional per monitor DPI awareness

### PR DESCRIPTION
**Pull request by tjachmann**
For win32 adds per monitor DPI arewareness, so the window's DPI gets updated when it is moved to a monitor with a different DPI. 
 - Reverts to system DPI, if an older windows then Windows 10 Creators Update is used
 - Although I think this behaviour should be the default, I made it optional for now.  See flag in backend-specific options in app_desc

### CODE REVIEW

File: sokol_app.h [hunk 1]
```diff
@@ -6458,6 +6462,35 @@ _SOKOL_PRIVATE LRESULT CALLBACK _sapp_win32_wndproc(HWND hWnd, UINT uMsg, WPARAM
                     }
                 }
                 break;
+
```
Comment (floooh) [inline sokol_app.h:6465]: please remove 1 line of whitespace

File: sokol_app.h [hunk 2]
```diff
@@ -6458,6 +6462,35 @@ _SOKOL_PRIVATE LRESULT CALLBACK _sapp_win32_wndproc(HWND hWnd, UINT uMsg, WPARAM
...
+                        _sapp.win32.dpi.window_scale = (float)dpix / 96.0f;
+                        _sapp.win32.dpi.content_scale = _sapp.win32.dpi.window_scale;
+                        _sapp.dpi_scale = _sapp.win32.dpi.window_scale;
+
+                        LPRECT lpRect = (LPRECT)lParam;
+                        SetWindowPos(hWnd, nullptr, lpRect->left, lpRect->top, lpRect->right - lpRect->left, lpRect->bottom - lpRect->top, SWP_NOZORDER | SWP_NOACTIVATE);
+                    }
+
+                    FreeLibrary(user32);
+                }
```
Comment (floooh) [inline sokol_app.h:6489]: I think it would be better to wrap this code into a small helper function _sapp_win32_dpi_changed().

File: sokol_app.h [hunk 3]
```diff
@@ -6698,37 +6736,49 @@ _SOKOL_PRIVATE void _sapp_win32_restore_console(void) {
 }
 
 _SOKOL_PRIVATE void _sapp_win32_init_dpi(void) {
-
+    
+    DECLARE_HANDLE(DPI_AWARENESS_CONTEXT_T);
```
Comment (floooh) [inline sokol_app.h:6740]: Should this be moved up into:

https://github.com/floooh/sokol/blob/b2f60754841c0bbfad47da86974e43e494f11bef/sokol_app.h#L2114-L2127

File: sokol_app.h [hunk 4]
```diff
@@ -6698,37 +6736,49 @@ _SOKOL_PRIVATE void _sapp_win32_restore_console(void) {
...
+            if (!(_sapp.desc.win32_per_monitor_dpi_v2 && fn_setprocessdpiawarenesscontext&&fn_setprocessdpiawarenesscontext((DPI_AWARENESS_CONTEXT_T)-4)))
+                fn_setprocessdpiawareness(process_dpi_awareness); // fallback system dpi aware
         }
-        fn_setprocessdpiawareness(process_dpi_awareness);
     }
     else if (fn_setprocessdpiaware) {
         fn_setprocessdpiaware();
         _sapp.win32.dpi.aware = true;
     }
+
```
Comment (floooh) [inline sokol_app.h:6781]: whitespace

File: sokol_app.h [hunk 5]
```diff
@@ -6738,6 +6788,7 @@ _SOKOL_PRIVATE void _sapp_win32_init_dpi(void) {
         _SOKOL_UNUSED(hr);
         SOKOL_ASSERT(SUCCEEDED(hr));
         /* clamp window scale to an integer factor */
+
```
Comment (floooh) [inline sokol_app.h:6791]: whitespace

File: sokol_app.h [hunk 6]
```diff
@@ -6752,6 +6803,7 @@ _SOKOL_PRIVATE void _sapp_win32_init_dpi(void) {
         _sapp.win32.dpi.mouse_scale = 1.0f / _sapp.win32.dpi.window_scale;
     }
     _sapp.dpi_scale = _sapp.win32.dpi.content_scale;
+
```
Comment (floooh) [inline sokol_app.h:6806]: whitespace

**Review by floooh — CHANGES_REQUESTED** [review]

**Comment by floooh** [comment]
Looks good!

Sorry about all the "remove whitespace" requests.

The new ```win32_default_size``` config setting is unrelated to the DPI changes though right?

**Comment by floooh** [comment]
...I've been sleeping over the default window size thing, and I think it's better to turn this into a cross-platform-thing by simply providing zero-initialized width and height  in sapp_desc. I'll check if the other desktop platforms also provide some sort of default-window-size feature (on web and mobile, the size is ignored anyway).

You don't need to change the PR for this, I'll take care of that during merging.

I'll first take care of https://github.com/floooh/sokol/pull/611 and then start integrating your PR (with the default-size changes).

**Comment by floooh** [comment]
Looks like you are compiling as C++? There's a ```nullptr``` in the call to SetWindowPos ;)

I'll clean up that stuff in a merge branch, I'll probably also move the LoadLibrary() and GetProcAddress() call out of the dpi-changed code path into the initialization code.

Also I didn't notice the -4 constant at first here:

```
fn_setprocessdpiawarenesscontext((DPI_AWARENESS_CONTEXT_T)-4)
```

Is there a human readable name for that constant?

**Comment by floooh** [comment]
Ok, a question:

The per-monitor DPI code-path is currently enabled with ```sapp_desc.win32_per_monitor_dpi_v2```. Is there a situation where one wouldn't want to set this to true? If not I would remove this config setting.

**Comment by floooh** [comment]
Here's my attempt to somewhat simplify the DPI setup code in _sapp_win32_dpi_init(), does that look alright to you?

```c
    /*
        NOTE on SetProcessDpiAware() vs SetProcessDpiAwareness() vs SetProcessDpiAwarenessContext():

        These are different attempts to get DPI handling on Windows right, from oldest
        to newest. SetProcessDpiAwarenessContext() is required for the new
        DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2 method.
    */
    if (fn_setprocessdpiawareness) {
        if (_sapp.desc.high_dpi) {
            /* app requests HighDPI rendering, first try the Win10 Creator Update per-monitor-dpi awareness,
               if that fails, fall back to system-dpi-awareness
            */
            _sapp.win32.dpi.aware = true;
            DPI_AWARENESS_CONTEXT_T per_monitor_aware_v2 = (DPI_AWARENESS_CONTEXT_T)-4;
            if (!(fn_setprocessdpiawarenesscontext && fn_setprocessdpiawarenesscontext(per_monitor_aware_v2))) {
                // fallback to system-dpi-aware
                fn_setprocessdpiawareness(PROCESS_SYSTEM_DPI_AWARE);
            }
        }
        else {
            /* if the app didn't request HighDPI rendering, let Windows do the upscaling */
            _sapp.win32.dpi.aware = false;
            fn_setprocessdpiawareness(PROCESS_DPI_UNAWARE);
        }
    }
    else if (fn_setprocessdpiaware) {
        // fallback for Windows 7
        _sapp.win32.dpi.aware = true;
        fn_setprocessdpiaware();
    }
```

**Comment by floooh** [comment]
FYI: this is the temporary merge branch I'm currently working on:

https://github.com/floooh/sokol/compare/tjachmann-master

**Comment by tjachmann** [comment]
> Looks like you are compiling as C++? There's a `nullptr` in the call to SetWindowPos ;)

 Yep and ops :) 

> fn_setprocessdpiawarenesscontext((DPI_AWARENESS_CONTEXT_T)-4)

You could alternativly replace the beginning of the _sapp_win32_init_dpi like this

#ifndef _DPI_AWARENESS_CONTEXTS_
    DECLARE_HANDLE(DPI_AWARENESS_CONTEXT);
#endif

#ifndef DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2
    #define DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2 (DPI_AWARENESS_CONTEXT)-4
#endif

typedef BOOL(WINAPI * SETPROCESSDPIAWARE_T)(void);
typedef bool (WINAPI * SETPROCESSDPIAWARENESSCONTEXT_T)(DPI_AWARENESS_CONTEXT); // since Windows 10, version 1703`


And then use the DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2 instead of -4. A bit longer, but somehow cleaner.

**Comment by tjachmann** [comment]
> Ok, a question:
> 
> The per-monitor DPI code-path is currently enabled with `sapp_desc.win32_per_monitor_dpi_v2`. Is there a situation where one wouldn't want to set this to true? If not I would remove this config setting.

I also think it should become the default. The only issue I could currently imagine is that when you pass in an concrete window size on creation, then the initial window creating does not respect per-monitor DPI.  More exactly if windows decides to create the window on another monitor the where you query the DPI.

**Comment by floooh** [comment]
I just did this, I think that's alright. It's just the -4 without context as function arg I had a problem with.

```c
DPI_AWARENESS_CONTEXT_T per_monitor_aware_v2 = (DPI_AWARENESS_CONTEXT_T)-4;
```

**Comment by floooh** [comment]
> More exactly if windows decides to create the window on another monitor the where you query the DPI.

Yep, but that's not supported currently anyway, and for this the code which initially queries the DPI would need to be changed anyway. I just removed that config item in the merge branch and I think it's working ok. So far I have only tested on a single-monitor setup by changing the DPI in the system settings, but next I'll check on my HighDPI laptop with a 1080p monitor attached...

**Comment by tjachmann** [comment]
> Here's my attempt to somewhat simplify the DPI setup code in _sapp_win32_dpi_init(), does that look alright to you?
> 
> ```c
>     /*
>         NOTE on SetProcessDpiAware() vs SetProcessDpiAwareness() vs SetProcessDpiAwarenessContext():
> 
>         These are different attempts to get DPI handling on Windows right, from oldest
>         to newest. SetProcessDpiAwarenessContext() is required for the new
>         DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2 method.
>     */
>     if (fn_setprocessdpiawareness) {
>         if (_sapp.desc.high_dpi) {
>             /* app requests HighDPI rendering, first try the Win10 Creator Update per-monitor-dpi awareness,
>                if that fails, fall back to system-dpi-awareness
>             */
>             _sapp.win32.dpi.aware = true;
>             DPI_AWARENESS_CONTEXT_T per_monitor_aware_v2 = (DPI_AWARENESS_CONTEXT_T)-4;
>             if (!(fn_setprocessdpiawarenesscontext && fn_setprocessdpiawarenesscontext(per_monitor_aware_v2))) {
>                 // fallback to system-dpi-aware
>                 fn_setprocessdpiawareness(PROCESS_SYSTEM_DPI_AWARE);
>             }
>         }
>         else {
>             /* if the app didn't request HighDPI rendering, let Windows do the upscaling */
>             _sapp.win32.dpi.aware = false;
>             fn_setprocessdpiawareness(PROCESS_DPI_UNAWARE);
>         }
>     }
>     else if (fn_setprocessdpiaware) {
>         // fallback for Windows 7
>         _sapp.win32.dpi.aware = true;
>         fn_setprocessdpiaware();
>     }
> ```

Looks good

**Comment by floooh** [comment]
Hmm, I need to fix sokol_imgui.h somehow, the scaling is now too big when moving from a HighDPI to a LowDPI monitor. I guess there are some places where the DPI scale factor isn't expected to change.

Interestingly, with the old SetProcessDPIAwareness() it worked as expected. I'll continue working on this stuff for the next few days I think.

**Comment by floooh** [comment]
...looks like image upload is currently borked... I can't even fix the previous comment hmm...

**Comment by floooh** [comment]
...while at it I also implemented per-monitor DPI for macOS (together with a generally more robust dpi_scale handling code).

Also reminder to self: check if on Windows an SAPP_EVENTTYPE_RESIZED is issued when the DPI changes (it should, as a "delayed side effect" of the SetWindowPos() call.

**Comment by tjachmann** [comment]
Ah to compile with legacy SDKs (WINVER < 0x0601), need to check if the DPI message is defined...

#ifndef WM_DPICHANGED
    #define WM_DPICHANGED                   0x02E0
#endif

**Comment by floooh** [comment]
> Ah to compile with legacy SDKs (WINVER < 0x0601), need to check if the DPI message is defined...

Good point, I'll add that. The merge will most likely happen tomorrow btw. Got a bit tangled up with fixing the new timing code when moving the window between displays with a different refresh rate.

**Comment by floooh** [comment]
Ok merged. Changelog entry (18-Jan-2022): https://github.com/floooh/sokol/blob/master/CHANGELOG.md

**Comment by floooh** [comment]
PS: as I mentioned above, I haven't included the default-size handling in this PR. I will come up with a different solution (after looking into the sokol_audio.h sample format problem).

**Comment by floooh** [comment]
@tjachmann FYI, the default-window-size changes are now implemented (see: https://github.com/floooh/sokol/pull/623)

**Pull request merged**